Tasks/Maven/maventask.ts

```diff
@@ -3,97 +3,181 @@
+    process.exit(1);
+})
+.then(function (code) {
+    // Setup tool runner to execute Maven goals
+    var mvnRun = tl.createToolRunner(mvnExec);
+    mvnRun.arg('-f');
+    mvnRun.pathArg(mavenPOMFile);
+    mvnRun.argString(mavenOptions);
+    mvnRun.arg(mavenGoals);
+    
+    // Read Maven standard output
+    mvnRun.on('stdout', function (data) {
+        processMavenOutput(data);
+    });
+    
+    // 2. Run Maven with the user goals. Compilation or test errors will cause this to fail.
+    return mvnRun.exec(); // Run Maven with the user specified goals
+})
+.fail(function (err) {
+    console.error(err.message);
+    userRunFailed = true; // Record the error and continue
+})
+.then(function (code) {
+    // 3. Always try to run the SonarQube analysis if it is enabled.
+    var mvnsq = getSonarQubeRunner();
+    if (mvnsq) {
+        // Run Maven with the sonar:sonar goal, even if the user-goal Maven failed (e.g. test failures).
+        // Note that running sonar:sonar along with the user goals is not supported due to a SonarQube bug.
+        return mvnsq.exec()
+    }
+})
+.fail(function (err) {
+    console.error(err.message);
+    console.error("SonarQube analysis failed");
+    sonarQubeRunFailed = true;
+})
+.then(function () {
+    // 4. Always publish test results even if tests fail, causing this task to fail.
     if (publishJUnitResults == 'true') {
-        //check for pattern in testResultsFiles
-        if (testResultsFiles.indexOf('*') >= 0 || testResultsFiles.indexOf('?') >= 0) {
-            tl.debug('Pattern found in testResultsFiles parameter');
-            var buildFolder = tl.getVariable('agent.buildDirectory');
-            tl.debug(`buildFolder=${buildFolder}`);
-            var allFiles = tl.find(buildFolder);
-            matchingTestResultsFiles = tl.match(allFiles, testResultsFiles, {
-                matchBase: true
-            });
-        } else {
-            tl.debug('No pattern found in testResultsFiles parameter');
-            matchingTestResultsFiles = [testResultsFiles];
-        }
+        publishTestResults(testResultsFiles);
+    }
+    
+    // Set overall success or failure
+    if (userRunFailed || sonarQubeRunFailed) {
+        tl.exit(1); // Set task failure
+    }
+    else {
+        tl.exit(0); // Set task success
+    }
 
-        if (!matchingTestResultsFiles || matchingTestResultsFiles.length == 0) {
-            tl.warning('No test result files matching ' + testResultsFiles + ' were found, so publishing JUnit test results is being skipped.');
-            return 0;
-        }
-        
-        var tp = new tl.TestPublisher("JUnit");
-        tp.publish(matchingTestResultsFiles, true, "", "", "", true);
+    // Do not force an exit as publishing results is async and it won't have finished 
+})
+
+// Publishes test results from files matching the specified pattern.
+function publishTestResults(testResultsFiles: string) {
```

Comment:
This function is junit specific now, maybe consider rename the function to make it apparent - publishJUnitTestResults?